Add Goal Tracking Infobox panel

BlueSoapTurtle opened the pull request:
This change adds new infoboxes to track progress towards a goal and adds a config section for picking this and changing some settings.
~~This is based on top of #15 which needs to be merged first, idk how github will handle that but we'll see!~~

### Changes
- **Goal Tracking Infobox:** Displays the overall percentage towards the selected reward goal, with individual component progress visible on hover.
- **Resin Tracking Infoboxes:** Shows individual resin type progress towards the goal. Configurable to show either the percentage completed or the remaining amount needed, capped at 100% and 0% as applicable.
- Adds a new method for checking if the player is in the lab region which includes the area just outside the actual lab
- Reward Tracking Config section:
  - **Selected Reward:** Allows users to select a reward to track resin progress towards.
  - **Goal Tracking Infobox:** Toggles the infobox displaying progress towards the selected reward.
  - **Resin Infoboxes:** Enables tracking infoboxes for each resin type.
  - **Resin Infobox Display Type:** Configures the resin infoboxes to display either the percentage completed or the amount left for each resin type.

### Images
The goal tracker
![image](https://github.com/user-attachments/assets/162fd48f-623f-4cb8-8618-1f5410c97d95)
The config
![image](https://github.com/user-attachments/assets/2e88c6c3-4697-4965-aab2-bf43643b3886)
Resin trackers showing how much left till goal
![image](https://github.com/user-attachments/assets/cac47abd-0202-4802-b8c4-81268f4df872)
Resin trackers showing percentage completed
![image](https://github.com/user-attachments/assets/f2fa710f-1632-438d-a3ef-3a8c744e3455)

Let me know if anything needs changing :)

Notloc commented:
Looking at this image, I find it a bit hard to read the numbers due to the spacing. They kinda mash together.
![17302704900741676982320360271996](https://github.com/user-attachments/assets/b07564e1-307a-457b-b510-09686c7d10dd)

Perhaps when at 1000 and above we could switch to k? 
i.e 1250 shows as 1.2k
I feel like that might improve the legibility.

BlueSoapTurtle commented:
> Looking at this image, I find it a bit hard to read the numbers due to the spacing. They kinda mash together. ![17302704900741676982320360271996](https://private-user-images.githubusercontent.com/23246298/381422212-b07564e1-307a-457b-b510-09686c7d10dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNzQzNjksIm5iZiI6MTczMDI3NDA2OSwicGF0aCI6Ii8yMzI0NjI5OC8zODE0MjIyMTItYjA3NTY0ZTEtMzA3YS00NTdiLWI1MTAtMDk2ODZjN2QxMGRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDA3NDEwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3YWQ0MWNlNTY5NjM1N2Y2ZDhhMjczNjE0ODFiOWQ1ZjVkOWVmYTY1YzAzZjE0NjE5ZDVmMDU2ODBjM2I1ZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3C7rCH6vrZZPM-bsdGjnw4I3Zm4lhLwJRKyjwub8s7s)
> 
> Perhaps when at 1000 and above we could switch to k? i.e 1250 shows as 1.2k I feel like that might improve the legibility.

That might be nice yeah! I'm using this atm to format the number
QuantityFormatter.quantityToStackSize()
It changes to k etc over 10k but will see if I can make it kick in at lower values

BlueSoapTurtle commented:
Ok, did a bit of work on this and came up with this:
![image](https://github.com/user-attachments/assets/a717e599-9128-4134-a5b9-7a2b012ebe23)
![image](https://github.com/user-attachments/assets/ea117837-a18b-435d-b02b-33d67be796be)
![image](https://github.com/user-attachments/assets/f4602ca4-6e84-4da4-ac89-c19c9447ceea)

Much cleaner than the other infoboxes <3

some questions I need feedback on:
- Do I just remove the infoboxes in favor of this? (I vote yes)
- Should I add an option to disable the bars? just a simple toggle would do it
- Should I change the bars when they hit 100%? ex make them gray or darker,
- Do we want percentage to go above 100?

Notloc commented:
Wow, that looks brilliant!
* I vote for the removal of the info boxes, its a very cluttered ui element these days.
* Probably a good option, maybe flavor it as a compact display option
* As for full bar styling, here are some quick rough visualizations of some ideas:
![image](https://github.com/user-attachments/assets/27ec3316-4126-437b-a063-bb342d771c98)
![image](https://github.com/user-attachments/assets/8270dce8-fccc-4a14-aaf4-29fd980279f2)
![image](https://github.com/user-attachments/assets/fc62c55e-953a-4a3a-8b43-cc1bcffc3ff5)
![image](https://github.com/user-attachments/assets/1795e121-1c1b-4eb4-9771-63b8864477f4)

BlueSoapTurtle commented:
Nice @Notloc ! I tried out a few things as well but it's pretty visible when the bar is full, ex:
making the bar gray
![image](https://github.com/user-attachments/assets/7b463732-895c-4925-91ea-81031d6c9f32)
or removing each bar
![image](https://github.com/user-attachments/assets/b5f75a07-4452-4b1b-ade5-3180d75da52f)

Probably fine to keep it as is, there's also no nice way to add a border to the bars sadly, here's the config:
![image](https://github.com/user-attachments/assets/96ec1eca-878e-468a-a8b9-b0b245bfee33)
you can pick None to disable the infobox
![image](https://github.com/user-attachments/assets/654f81b3-3222-4250-a78b-72d68c82bb2d)

Lmk if you have some suggestions or changes

Notloc commented:
My only feedback so far is that the default position and size should probably be adjusted.
The default is too narrow and behind the mini-game ui.

Notloc commented:
Specifically for the Aldarium goal, would it be possible to add a configurable multiplier?
Ends up a bit useless when it only tracks one at a time, and long term, for me at least, its gonna be the main reason I'm playing this mini-game.
![image](https://github.com/user-attachments/assets/589352e0-825a-4cf9-b889-a3f40044b4be)


Other than that, this is fantastic! The way the bars provide both context to the resin numbers and visual feedback about if my potion making strategy is balanced is just fantastic.

BlueSoapTurtle commented:
That's a great idea @Notloc Thanks for the feedback! 
I agree, it's really nice to use this when doing the minigame as you say you know what to go for with the bars. 

Which is better of these?
Bold to the right
![image](https://github.com/user-attachments/assets/9e4ff443-6d1c-4be7-bb11-53e5e6e15cd3)
Normal to the right
![image](https://github.com/user-attachments/assets/c5e11a5e-6391-4d29-b915-205e4fa52c1a)

Alternatively, if they change the amount and the item is a repeatable reward we display the progress in numbers instead?
![image](https://github.com/user-attachments/assets/30c7712f-f680-4bba-a44d-86cb78b97fed)
Or maybe better to put that in the top part, like this:
![image](https://github.com/user-attachments/assets/3554b762-0ed2-4f4a-ac3b-8b789e6bc176)

Notloc commented:
I think the last one is nice, seeing both the percentage and the amount you can afford.

BlueSoapTurtle commented:
Perfect, that's fixed now then! Also fixed the position and size, ish. not sure why it's not working after I change the reward item:
![image](https://github.com/user-attachments/assets/9b142336-e60c-4272-89e7-98361fc6221d)
In the constructor I'm setting this:
```java
setPosition(OverlayPosition.TOP_RIGHT);
setPriority(PRIORITY_MED);
setPreferredSize(new Dimension(350, 0));
panelComponent.setPreferredSize(new Dimension(350, 0));
```

Changes:
- You can now change the reward quantity for repeatable rewards ex Aldarium
- Overlay GoalInfoBoxOverlay:
  - Repeatable rewards now show the current and goal amount to the right of the item name
  - Fixed position and size of the panel
  - Improved the code readability by creating the ComponentData in a loop and providing the 
    constructor with the neccessary fields, then using streams to calculate min affordable value and average percentage
- Changes to PotionComponent:
  - Reordered the PotionComponent enum to be MOX, AGA, LYE instead of AGA, LYE, MOX 
  - Changed PotionComponent to calculate the Color in the constructor instead of having to decode it all the time
  - Added resin varp field to be able to fetch the resin amount from a component more easily
  
  S
  Lmk if anyone has some comments to add to it

hex-agon commented on src/main/java/work/fking/masteringmixology/GoalInfoBoxOverlay.java:
```diff
@@ -0,0 +1,228 @@
+    private int rewardQuantity = 1;
+
+    private final Map<Integer, BufferedImage> rewardIconCache = new HashMap<>();
+    private final EnumMap<PotionComponent, BufferedImage> componentSpriteCache = new EnumMap<>(PotionComponent.class);
```
You can just cache these on individual vars when you construct the overlay

BlueSoapTurtle replied:
Alright! Got a bit busy due to moving but fixed it now. Restored caching on the reward icon and moved the component caching to individual variables

BlueSoapTurtle commented on src/main/java/work/fking/masteringmixology/MasteringMixologyPlugin.java:
```diff
@@ -129,6 +135,16 @@ public boolean isInLab() {
+    public boolean isInLabRegion() {
+        Player player = client.getLocalPlayer();
+        return player != null && player.getWorldLocation().getRegionID() == LABS_REGION_ID
+                && player.getWorldLocation().getPlane() == LABS_REGION_PLANE;
```
As stated in the comment above the method:
/**
 * @return true if the player is in the labs region (the area where the minigame takes place)
 * the isInlab method only checks if they are inside the actual lab room where the UI is active
 */
 This method is different as it also shows the goal tracker when the player is near the bank and npc that sells items, found this useful when going to the bank to see what I needed more of

BlueSoapTurtle commented on src/main/java/work/fking/masteringmixology/RewardItem.java:
```diff
@@ -0,0 +1,52 @@
+    private final String itemName;
+    private final int itemId;
+    private final boolean repeatable;
+    private final Map<PotionComponent, Integer> componentCost = new HashMap<>();
```
Done!

BlueSoapTurtle commented on src/main/java/work/fking/masteringmixology/GoalInfoBoxOverlay.java:
```diff
@@ -0,0 +1,228 @@
+    private final PanelComponent topPanel = new PanelComponent();
+
+    private final EnumMap<PotionComponent, ComponentData> componentDataMap = new EnumMap<>(PotionComponent.class);
+    private double overallProgress = 0.0;
+    private int itemsAffordable = 0;
+    private RewardItem rewardItem;
+    private int rewardQuantity = 1;
```
Done

BlueSoapTurtle commented on src/main/java/work/fking/masteringmixology/GoalInfoBoxOverlay.java:
```diff
@@ -0,0 +1,228 @@
+    @Override
+    public Dimension render(Graphics2D graphics) {
+        // Recalculate the data before checking if the overlay should be displayed
+        if (dataDirty) {
```
Done!

BlueSoapTurtle commented on src/main/java/work/fking/masteringmixology/GoalInfoBoxOverlay.java:
```diff
@@ -0,0 +1,228 @@
+            goalAmountText = QuantityFormatter.quantityToStackSize(itemsAffordable) + "/" + QuantityFormatter.quantityToStackSize(rewardQuantity);
+        }
+
+        final LineComponent topLine = LineComponent.builder()
```
Done

BlueSoapTurtle commented:
Lmk if there's anything else to fix! Otherwise ready for merge/review

hex-agon commented:
At first glance this seems ok, i'll play a bit with it on the weekend

hex-agon commented:
Merged with c6781dda2900f6ef896b577e57580e9f245b5f7e
